src/sspi/pku2u.rs

```diff
@@ -0,0 +1,797 @@
+    // we need it for the further authenticator checksum calculation
+    // https://datatracker.ietf.org/doc/html/draft-zhu-pku2u-04#section-6
+    // The checksum is performed on all previous NEGOEX messages in the context negotiation.
+    gss_api_messages: Vec<u8>,
+    negoex_random: [u8; RANDOM_ARRAY_SIZE],
+}
+
+impl Pku2u {
+    pub fn new_server_from_config(config: Pku2uConfig) -> Result<Self> {
+        let mut rng = OsRng::default();
+
+        Ok(Self {
+            mode: Pku2uMode::Server,
+            config,
+            state: Pku2uState::Preauthentication,
+            encryption_params: EncryptionParams::default_for_server(),
+            auth_identity: None,
+            conversation_id: Uuid::default(),
+            auth_scheme: Some(Uuid::from_str(DEFAULT_NEGOEX_AUTH_SCHEME).unwrap()),
+            seq_number: 2,
+            // https://www.rfc-editor.org/rfc/rfc4556.html#section-3.2.3
+            // Contains the nonce in the pkAuthenticator field in the request if the DH keys are NOT reused,
+            // 0 otherwise.
+            // generate dh parameters at the start in order to not waste time during authorization
+            dh_parameters: generate_server_dh_parameters(&mut rng)?,
+            negoex_messages: Vec::new(),
+            gss_api_messages: Vec::new(),
+            negoex_random: rng.gen::<[u8; RANDOM_ARRAY_SIZE]>(),
+        })
+    }
+
+    pub fn new_client_from_config(config: Pku2uConfig) -> Result<Self> {
+        let mut rng = OsRng::default();
+
+        Ok(Self {
+            mode: Pku2uMode::Client,
+            config,
+            state: Pku2uState::Negotiate,
+            encryption_params: EncryptionParams::default_for_client(),
+            auth_identity: None,
+            conversation_id: Uuid::new_v4(),
+            auth_scheme: None,
+            seq_number: 0,
+            // https://www.rfc-editor.org/rfc/rfc4556.html#section-3.2.3
+            // Contains the nonce in the pkAuthenticator field in the request if the DH keys are NOT reused,
+            // 0 otherwise.
+            // generate dh parameters at the start in order to not waste time during authorization
+            dh_parameters: generate_client_dh_parameters(&mut rng)?,
+            negoex_messages: Vec::new(),
+            gss_api_messages: Vec::new(),
+            negoex_random: rng.gen::<[u8; RANDOM_ARRAY_SIZE]>(),
+        })
+    }
+
+    pub fn next_seq_number(&mut self) -> u32 {
+        let seq_num = self.seq_number;
+        self.seq_number += 1;
+
+        seq_num
+    }
+}
+
+impl Sspi for Pku2u {
+    fn complete_auth_token(&mut self, _token: &mut [SecurityBuffer]) -> Result<SecurityStatus> {
+        Ok(SecurityStatus::Ok)
+    }
+
+    fn encrypt_message(
+        &mut self,
+        _flags: EncryptionFlags,
+        message: &mut [SecurityBuffer],
+        sequence_number: u32,
+    ) -> Result<SecurityStatus> {
+        SecurityBuffer::find_buffer_mut(message, SecurityBufferType::Token)?;
```

Comment:
The return value is not used so I suppose this is a sanity check making sure a "Token" buffer is in the message, am I right? I think it would help to have a comment explaining why this is necessary. Also, it’s not clear to me why the `_mut` variant is used.

Reply:
yes, it's just for checking that the Token buffer is just present. I added the comment about it and changed the function used to the not `_mut` variant